lru/lru.go

```diff
@@ -65,9 +75,13 @@ func (c *Cache) Add(key Key, value interface{}, size int64) bool {
 	}
 
 	if ee, ok := c.cache[key]; ok {
-		c.ll.MoveToFront(ee)
-		ee.Value.(*entry).value = value
-		return true
+		kv := ee.Value.(*entry)
+		if !kv.Expired() {
```

Comment:
this seems double work. Why not add below

c.RemoveExpired()
for c.Size > c.MaxSize {
     c.RemoveOldest()
}

Reply:
Because it's faster to only remove enough items to get Size below MaxSize. Doing a full pass over all items for every Add seems inefficient.

Reply:
The only way I can see to simplify this would change existing behavior. That is, don't update existing cache items on Add, always replace them with a fresh cache item.

Reply:
good point.

How about

for c.Size > c.MaxSize {
   c.RemoveExpired()
}
for c.Size > c.MaxSize {
    c.RemoveOldest()
}